graphistry/embed_utils.py

```diff
@@ -0,0 +1,593 @@
+            use_feat (bool, optional): whether to featurize nodes, if False
+                will produce random embeddings and shape them during training.
+                Defaults to True.
+            X (List or pd.DataFrame, optional): Which columns in the nodes
+                dataframe to featurize. Inherets args from
+                graphistry.featurize()
+                Defaults to None.
+            lr (float, optional): learning rate. Defaults to 0.003.
+            evaluate (bool, optional): whether to evaluate the model.
+                Defaults to False.
+            epochs (int, optional): traing epoch. Defaults to 2.
+            batch_size (int, optional): batch size. Defaults to 32.
+            train_split (float, optional): train percentage, between 0, 1.
+                Defaults to 0.8.
+
+        Returns:
+            self: graphistry instance
+        """
+        if inplace:
+            res = self
+        else:
+            res = self.bind()
+
+        res._relation = relation
+        res._use_feat = use_feat
+        res._embed_dim = embedding_dim
+        res._train_split = train_split
+        res._eval_flag = evaluate
+
+        if callable(proto):
+            res.proto = proto
+        else:
+            res.proto = res.protocol[proto]
+
+        if res._use_feat and res._nodes is not None:
+            # todo decouple self from res
+            res = res.featurize(kind="nodes", X=X, *args, **kwargs)
+
+        if not hasattr(self, 'triplets'):
+            res._preprocess_embedding_data(train_split=train_split)
+            res._build_graph()
+
+        return res._train_embedding(epochs, batch_size, lr=lr, device=device)
+
+    def calculate_prob(
+        self,
+        test_triplet,
+        test_triplets,
+        threshold,
+        h_r,
+        node_embeddings,
+        infer=None
+    ):
+        # TODO: simplify
+        if infer == "all":
+            s, r, o_ = test_triplet
+        else:
+            s, r = test_triplet
+
+        subject_relation = test_triplet[:2]
+        num_entity = len(node_embeddings)
+
+        delete_idx = torch.sum(h_r == subject_relation, dim=1)
+        delete_idx = torch.nonzero(delete_idx == 2).squeeze()
+        delete_entity_idx = test_triplets[delete_idx, 2].view(-1).numpy()
+        perturb_entity_idx = np.array(
+                list(set(np.arange(num_entity)) - set(delete_entity_idx))
+        )
+        perturb_entity_idx = torch.from_numpy(perturb_entity_idx).squeeze()
+
+        if infer == "all":
+            perturb_entity_idx = torch.cat(
+                    (perturb_entity_idx, torch.unsqueeze(o_, 0))
+            )
+
+        o = self.proto(
+                node_embeddings[s],
+                self._embed_model.relational_embedding[r],
+                node_embeddings[perturb_entity_idx])
+
+        score = torch.sigmoid(o)
+        return perturb_entity_idx[score > threshold]
+
+    def _predict(
```

Comment:
docstr

we need a clear public method (and docstr) for, given a trained embedding, to embed a new graph, incl how to handle feature engineering

Reply:
we have one, its `predict_link` which predicts d for given s and r

Reply:
- predicting a d for (s,r) makes sense is recsys
- scoring (s,r,d) seems diff, and more relevant to anomaly detection in fraud/cyber, e.g., bottom of https://github.com/graphistry/pygraphistry/blob/heteroembed/demos/more_examples/graphistry_features/embed/simple-ssh-logs-rgcn-anomaly-detector.ipynb